pygmt/src/triangulate.py

```diff
@@ -0,0 +1,384 @@
+        finds how the points should be connected to give the most equilateral
+        triangulation possible. If a map projection (give *region* and
+        *projection*) is chosen then it is applied before the triangulation is
+        calculated. The actual algorithm used in the triangulations is either
+        that of Watson [1982] [Default] or Shewchuk [1996] (if installed; type
+        **gmt triangulate -** on the command line to see which method is
+        selected). This choice is made during the GMT installation.
+        Furthermore, if the Shewchuk algorithm is installed then you can also
+        perform the calculation of Voronoi polygons and optionally grid your
+        data via the natural nearest neighbor algorithm. **Note**: For
+        geographic data with global or very large extent you should consider
+        :gmt-docs:`sphtriangulate <sphtriangulate.html>` instead since
+        **triangulate** is a Cartesian or small-geographic area operator and is
+        unaware of periodic or polar boundary conditions.
+
+        Must provide either ``data`` or ``x``, ``y``, and ``z``.
+
+        Full option list at :gmt-docs:`triangulate.html`
+
+        Parameters
+        ----------
+        x/y/z : np.ndarray
+            Arrays of x and y coordinates and values z of the data points.
+        data : str or {table-like}
+            Pass in (x, y, z) or (longitude, latitude, elevation) values by
+            providing a file name to an ASCII data table, a 2D
+            {table-classes}.
+        {J}
+        {R}
+        outfile : str or bool or None
+            The name of the output ASCII file to store the results of the
+            histogram equalization in.
+        {V}
+        {b}
+        {d}
+        {e}
+        {f}
+        {h}
+        {i}
+        {s}
+        {w}
+
+        Returns
+        -------
+        ret: pandas.DataFrame or None
+            Return type depends on the ``outfile`` parameter:
+
+            - pandas.DataFrame if ``outfile`` is True or None
+            - None if ``outfile`` is a str (file output is stored in
+              ``outfile``)
+        """
+        # Return a pandas.DataFrame if ``outfile`` is not set
+        if output_type not in ["numpy", "pandas", "file"]:
+            raise GMTInvalidInput(
+                "Must specify 'output_type' either as 'numpy', 'pandas' or 'file'."
+            )
+
+        if isinstance(outfile, str) and output_type != "file":
+            msg = (
+                f"Changing 'output_type' from '{output_type}' to 'file' "
+                "since 'outfile' parameter is set. Please use output_type='file' "
+                "to silence this warning."
+            )
+            warnings.warn(message=msg, category=RuntimeWarning, stacklevel=2)
+            output_type = "file"
+
+        if header is not None and output_type != "file":
+            raise GMTInvalidInput("'header' is only allowed with output_type='file'.")
```

Comment:
This is more tricky than for grdhisteq because the input is tabular, so a user may need to use **-hi** with file input to specify what records should be taken as headers. One option would be to also check if the first character in the argument is "o".

Reply:
True, but someone could also do `header=["i2", "o+c"]`, `header="o+c"`, or stuff like that, and that would be too many if-statements to check.

Reply:
True. The issue that we're trying to avoid is having pandas.read_csv fail without giving any useful information about why it fails. Only a quick thought, but maybe we could use a try/except when reading the temporary file output rather than an if-statement depending on "header" before calling triangulate.

Reply:
Actually, how about:

```suggestion
        if header is not None and "o" in str(header) and output_type != "file":
            raise GMTInvalidInput("'header' is only allowed with output_type='file'.")
```

Reply:
Hmm, I actually cannot seem to figure out how **-ho** affects the output at all for triangulate (or select). For grd2xyz, only **-ho+c** works. I agree that it would be fine to remove it, but there may be an upstream issue here that will require updates later on. I'll look into it more after the PyGMT release.

Reply:
Ok, done at ed0ba2c30a7a86871537ef663b06e48c4c184c93, will leave this thread open for now.